### PR TITLE
[SPARK-2673][CORE][WEBUI] Enable to attach Debugger to Executors easily

### PR DESCRIPTION
**Pull request by sarutak**
In current implementation, it's difficult to attach debugger to each Executor in the cluster.
There are some reasons as follows.

1) Multi Executors can run on the same machine so each executor should open individual debug ports.

2)  Even if we can open unique debug port to each Executors running on the same machine, it's a bother to check debug port of each executor.

To solve those problem, I think following 2 improvement is needed.

1) Enable executor to open unique debug port on a machine by setting debug port to 0.
2) Expand WebUI to be able to show debug ports opening in each executor.

Following image is how Web UI(ExecutorsPage) shows the debug ports for each Executor.

debug-port-feature

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala [hunk 1]
```diff
@@ -20,6 +20,8 @@ package org.apache.spark.executor
 import java.net.URL
 import java.nio.ByteBuffer
 
+import sun.misc.VMSupport
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: This is risky to use, even if it is in OpenJDK and Oracle's JVM.

Comment (sarutak) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: So, what about enabling when we build with a specific profile?

**Comment by srowen** [comment]
Referring to the JIRA, I do wonder how much this adds over just setting up the agent manually with JVM opts, because it touches a reasonable amount of code. Yes, you have to pick a fixed port, but then problem 2 goes away; you know what the agent port is. Yes, that doesn't work with several executors per machine, but if you're restarting your app anyway to debug, would you not perhaps just run a leaner config with 1 per machine? in the event that you really have to debug a multiple executor per machine job, maybe you can make the port a function of the executor's ID? I think we added some support for that?

**Comment by WangTaoTheTonic** [comment]
What about ApplicationMaster (ExecutorLauncher) and Driver side? As there may be too much executors so we attach debuggers to them for convienience, and there's no same trouble in AM and Driver, am I right?

**Comment by andrewor14** [comment]
@sarutak are you still actively working on this? If not, would you mind closing this patch for now? If there's interest in the future we can always reopen it against the latest master branch.

**Comment by sarutak** [comment]
Thanks for the notification. Unfortunately, now I don't have time to improve this patch. So I'll close for now.

**Pull request closed without merging**